src/openmc_heat_driver.cpp

```diff
@@ -226,7 +219,7 @@ void OpenmcHeatDriver::init_densities()
             mass += c.get_density() * vol;
           }
 
-          densities_(fluid_index + fluid_offset) = mass / total_vol;
+          densities_(fluid_index) = mass / total_vol;
         }
       }
```

Comment:
Would you consider something like this?  In your version, `i` and `j` aren't used to compute or subscript anything; and `fluid_index` is an unbounded subscript.  Below should be equivalent

```
gsl::index elem_idx = 0;
while (elem_idx < n_fluid) {
  
  double mass = 0.0;
  double total_vol = 0.0;
  
  for (const auto &cell_idx : elem_to_cell_inst_[elem_idx++]) {
    const auto &c = openmc_driver_->cells_[cell_idx]; 
    double vol = c.volume_;
    
    total_vol += vol;
    mass += c.get_density() * vol;
  }
  
  densities_(elem_idx) = mass / total_vol;
}

Reply:
There are also a couple things that confuse me about this, where you increment `elem_to_cell_inst_[fluid_index++]`

```
          for (const auto& idx : elem_to_cell_inst_[fluid_index++]) {
            const auto& c = openmc_driver_->cells_[idx];
            double vol = c.volume_;

            total_vol += vol;
            mass += c.get_density() * vol;
          }

          densities_(fluid_index) = mass / total_vol;
```

1. On the final iteration of the surrounding loops (`i == n_pins_-1`, `j==n_axial_-1`), when you reference `densities_(fluid_index)`, will `fluid_index` be equal to `n_pins_*n_axial_` at that point?  
2.  For a given `i`, do `elem_to_cell_inst_[i]` and `densities_(i)` refer to different elements?

Reply:
I agree that we don't need loops for `i` and `j`. Personally I would prefer to see a single `for` loop that increments `fluid_index` rather than a `while` loop.

Reply:
I'm also on board with collapsing the loops to a loop for `fluid_index`. Also, @RonRahaman I'm not sure why that increment is there... I was originally incrementing because `fluid_index` wasn't being set by the loop and I was offsetting, but I think there's been a bug here all along... I wonder why it didn't cause any seg faults though..

Reply:
For a given `i`, `elem_to_cell_inst_` provides a list of OpenMC cells that map to that T/H element, while `densities_` is the fluid density on the T/H elements.

Reply:
Now that the increment is gone, everything makes much more sense!  A for loop works, and the correspondence between `elem_to_cell_inst_` and `densities_` is clear.  Thanks!